### PR TITLE
[FEATURE] support for registry config in wasmcloud.toml 

### PR DESCRIPTION
**Pull request by ahmedtadde**
## Feature or Problem
As a user, I would really like it if I didn't have to manually type out my URL to push my actor every time I need to push a new version.

## Related Issues
#860 

## Release Information
`wash-cli v0.23.0`

## Consumer Impact
better DX

### Acceptance or Integration
[intergration_reg_config](https://github.com/ahmedtadde/wasmCloud/blob/63d95e548e5fa063bcc7c38ae4009c6f85928f72/crates/wash-cli/tests/wash_reg.rs#L248)

### CODE REVIEW

File: crates/wash-lib/src/cli/registry.rs
Comment (vados-cosmonic) [inline crates/wash-lib/src/cli/registry.rs:45]: Hmnnn I think the ENV variable might be a bit weirdly named. This URL is normally the URL of *the artifact*, which happens to be in a registry.  Maybe this should be `WASH_ARTIFACT_URL` ?

As written, I'd think of a registry URL to be `ghcr.io/user`, and an artifact URL to be `ghcr.io/user/project:0.1.0`... does that match with how you're thinking about it?

File: crates/wash-lib/src/parser/mod.rs
Comment (vados-cosmonic) [inline crates/wash-lib/src/parser/mod.rs:237]:
```suggestion
    /// Optional artifact OCI registry configuration. Primarily used for `wash push` & `wash pull` commands
```

File: crates/wash-cli/tests/wash_reg.rs
Comment (vados-cosmonic) [inline crates/wash-cli/tests/wash_reg.rs:255]: Thanks for adding to the test suite! These tests look great, but they're a bit focused on lower level bits -- I think they might be better as unit tests closer to the functions that you're testing.

Also, rather than pull a bunch of tests into one, let's split them out into little chunks -- testing precedence, with the field missing, etc can be their own test functions (in the configuration code, rather than here IMO), with their own `///` comments explaining what they do/what they test, if you're open to it.

**Review by vados-cosmonic — COMMENTED** [review]
Thanks for taking on such a meaty task @ahmedtadde , this is awesome. This is something that will improve the UX of `wash pull`and `wash push` immensely :rocket: .

I found a few minor things and stuff you might want to think about reorganizing, if you wouldn't mind taking a quick look!

File: crates/wash-lib/src/cli/registry.rs
Comment (ahmedtadde) [reply crates/wash-lib/src/cli/registry.rs:45]: so, the distinction did cross my mind. Though, looking at the related issue (and its discussion) and the related code, there is no apparent distinction between a registry url and artifact url (ie, a single registry providing multiple artifact repositories). if the distinction does matter (and it probably should), then there needs to be more refinement on the issue to address how we want to handle it across all config options (cli, env, wasmcloud file, custom and default credentials file).

File: crates/wash-cli/tests/common/mod.rs
Comment (ahmedtadde) [reply crates/wash-cli/tests/common/mod.rs:90]: is there are particular reason why to `tokio` async variant is wanting here?

Comment (vados-cosmonic) [reply crates/wash-cli/tests/common/mod.rs:90]: Hey ahmed I think you were trying to note that [`tokio::fs::write`](https://docs.rs/tokio/latest/tokio/fs/fn.write.html) is preferrable, that's totally right, please use that! that is the better option since it's in an `async fn`

File: crates/wash-lib/src/cli/registry.rs
Comment (vados-cosmonic) [reply crates/wash-lib/src/cli/registry.rs:45]: Hmnnn, so when I look around the codebase I mostly see uses of URLs in the context of artifacts -- the places I do see a "registry url" (like `oci.rs`), It's intended to be the URL of the registry itself (not an individual manifest or artifact). 

What really made me think that the name might be more geared was two things:

- Doc comment above there that mentions artifacts (my first thought was that maybe we need to update that doc comment, but on second thought it seems like an artifact URL)
- How we use [the `url` parameter on `RegistryPushCommand`](https://github.com/wasmCloud/wasmCloud/blob/main/crates/wash-cli/src/common/registry_cmd.rs#L33) -- we call it an `artifact_url` there.

I'm willing to be flexible here though, if you feel like `WASH_REG_URL` is the right name for this. It's not like we can't change it later, and I'd love to get your change merged sooner rather than later!

Comment (ahmedtadde) [reply crates/wash-lib/src/cli/registry.rs:45]: on the decision on how to move forward with this, i will defer to @thomastaylor312 since he is the one who originally created the issue #860. the solution i proposed (see discussion on the issue itself) isn't a good fit for having a common configured registry to which multiple artifact can be pushed/pulled.  so, we can either move forward with this for now or this PR is put on ice until an adequate solution is devised.

Comment (vados-cosmonic) [reply crates/wash-lib/src/cli/registry.rs:45]: Let's move forward with this for now! Given that it wasn't used before I think we've got some space -- I'll leave this open just in case Taylor wants to chime in.

**Comment by vados-cosmonic** [comment]
Hey @ahmedtadde there is one conflict, but please let me know when this is ready to re-review and I'll review and merge!

File: crates/control-interface/src/types.rs
Comment (connorsmith256) [inline crates/control-interface/src/types.rs:227]: Nit: this could be `default_value` to avoid the need to define a function

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:41]: Nit: Instead of `.map_err` where we call `anyhow!`, you should be able to call `.context`, since the error type is already any anyhow error

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:56]: Consider adding a `.context` here in case the parse error isn't very friendly

File: crates/wash-lib/src/parser/mod.rs
Comment (connorsmith256) [inline crates/wash-lib/src/parser/mod.rs:575]: Nit: is this clone necessary? And you can avoid the unwrap by combining this with the above block, like so:
```rust
let Some(credentials_file) = credentials_file else {
  return Ok(RegistryAuth::Anonymous);
};
```

You can repeat this structure a couple times below as well

File: crates/wash-lib/src/parser/mod.rs
Comment (connorsmith256) [inline crates/wash-lib/src/parser/mod.rs:589]: Consider printing/returning an error here, since this case — the file exists but is malformed — sounds like a user error

File: crates/wash-lib/src/config.rs
Comment (connorsmith256) [inline crates/wash-lib/src/config.rs:48]: Hmm, I think we might want to associate the creds file with a context, instead of having a global one. What do you think?

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:69]: When will `get_config` return an error? Would it be better to return that error instead of defaulting to anonymous auth?

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:78]: Nit: since we're checking the same condition twice, you could assign to `(user, password)` above this block

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:104]: It looks like we have this same logic used 3 times. Can we move it to a function?

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:130]: Same question here: since this is used 3 times, can we move it to a function?

File: crates/wash-lib/src/cli/registry.rs
Comment (connorsmith256) [inline crates/wash-lib/src/cli/registry.rs:70]: Since `artifact` is always required, doesn't it make more sense to keep this as a positional argument?

File: crates/wash-lib/src/cli/registry.rs
Comment (connorsmith256) [reply crates/wash-lib/src/cli/registry.rs:45]: Taylor is out for the next week, but I'll comment on his behalf. Up to now, this url was meant to be a specific artifact URL, e.g. `wasmcloud.azurecr.io/echo:0.3.8`, not just a registry url (e.g. `wasmcloud.azurecr.io`). 

Currently, I can do:

```
wash reg pull/push wasmcloud.azurecr.io/echo:0.3.8
```

With your changes, we allow users to set `registry` config in the project file. So that _might_ allow usage like:
```
wash reg pull/push echo:0.3.8
```

However, since the registry config is optional, we won't always know what the registry URL is. Another option is to change the command structure to be something like this:
```
wash reg pull/push --registry wasmcloud.azurecr.io echo:0.3.8
```

And then in cases where I've set up `registry` config in my `wasmcloud.toml`, I could omit the flag and just do:

```
wash reg pull/push echo:0.3.8
```

File: crates/wash-lib/src/parser/mod.rs
Comment (connorsmith256) [inline crates/wash-lib/src/parser/mod.rs:562]: Should this be `registry`, or `artifact_url`? I noticed all the uses of this function pass a full artifact URL, not just the name of the registry

**Review by connorsmith256 — CHANGES_REQUESTED** [review]

File: crates/control-interface/src/types.rs
Comment (ahmedtadde) [reply crates/control-interface/src/types.rs:227]: i think i tried to use `default_value` but i got an error. i know it works with `clap` but couldn't get it to work with `serde`. can you please paste in the annotation you have in mind

Comment (connorsmith256) [reply crates/control-interface/src/types.rs:227]: Ah, I was thinking of clap, not serde. Disregard this 🙂

**Comment by ahmedtadde** [comment]
hey @connorsmith256 

so assuming that a registry url is configured (via `WASH_REG_URL` or `wasmcloud.toml`)... how do we deterministically handle

(1) `wash reg pull/push wasmcloud.azurecr.io/echo:0.3.8` vs (2) `wash reg pull/push echo:0.3.8`

That is, how do we distinguish between a full artifact url (1) and a repository path (2) that is supposed to be prepended by a registry url.  I am thinking using `oci_distribution:Reference::try_from()` could work? if it returns an error, then we assume the input value is meant to be just the repository path instead of the full artifact url.

assuming that the above is handled... the current `push/pull` commands would basically stay the same except that we add an additional optional arg `--registry` which fallbacks to `WASH_REG_URL` or `project_config.common.registry.url` when available. if the registry url fallback values are not available, we return an error to the user. 

now, for the credentials, does the credentials file `project_config.common.registry.credentials` take precedence over the current context's credentials file? or vice versa? either way, we only return an error when the credentials file is specified, exists and parsing it from `json` into `RegistryCredentialMap` fails.

**Comment by connorsmith256** [comment]
@ahmedtadde here's my proposal, though I'd want to hear from others, including you, since this is a breaking change to wash, no matter which approach we take:
- `--registry` becomes an optional argument to `push/pull`, instead of being a positional arg
  - (or `--registry-url`? `-r` for shorthand would be nice too)
  - for reference, we use `--oci-registry`/`OCI_REGISTRY` for the host 
- in terms of precedence, we should check in this order, which favors explicit user intent over implicit config:
  - flag
  - env var
  - `wasmcloud.toml`
- if no registry URL is provided at all, I agree, we should print an error
- for credentials, the precedence should follow the same order, for consistency. I don't think we need a default registry credsfile. If you think we should have one, I would prefer it to be per-context, instead of global
- **BONUS**: the full artifact URL could be inferred, from a combination of the registry URL + name + version. The artifact name and version could be provided the same ways as the registry URL. That would enable `wash reg push` without any extra details, for the most common workflow, and commands like `wash reg push -v 0.3.8-rc.1` to override a specific parameter
  - this could be done in subsequent PR(s)

File: crates/control-interface/src/types.rs
Comment (connorsmith256) [inline crates/control-interface/src/types.rs:242]: Nit: I think this could be accomplished via a `TryFrom`

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (connorsmith256) [inline crates/wash-cli/src/common/registry_cmd.rs:204]: I'm not very familiar with the `oci_distribution` crate. When will this not be true?

File: crates/wash-cli/tests/wash_reg.rs
Comment (connorsmith256) [inline crates/wash-cli/tests/wash_reg.rs:248]: Nit: typo in integration

**Review by connorsmith256 — APPROVED** [review]
Sorry for the delay @ahmedtadde, I've had a lot on my plate

LGTM! Just a few optional nits. @vados-cosmonic could you handle getting this merged?

**Comment by vados-cosmonic** [comment]
Sure @connorsmith256 ! Hey @ahmedtadde let me know if you want me to merge this PR immediately, or if you want to address those nits -- I'm fine either way as we can always come back and make the small changes later -- thanks again for the awesome contribution!

File: crates/wash-cli/src/common/registry_cmd.rs
Comment (ahmedtadde) [reply crates/wash-cli/src/common/registry_cmd.rs:204]: based on our use case and the implementation of `image.whole()` from the `oci_distribution` crate, that predicate will be false when the user provides a repository path rather than a full artifact url.

Comment (ahmedtadde) [reply crates/wash-cli/src/common/registry_cmd.rs:205]: will mark this as resolved but you follow up on this if you think there should be more deliberation on this

**Comment by ahmedtadde** [comment]
hey @vados-cosmonic , i pushed a commit to address the nits from @connorsmith256. from my end, PR should be good to merge at this point.

**Pull request merged**
